pennylane/qnodes/passthru.py

```diff
@@ -0,0 +1,145 @@
+
+    The advantages of this approach is that the qfunc can be differentiated using its AD framework
+    without requiring a separate method for computing the Jacobian, and that the internals
+    of the simulation are visible in the computational graph.
+
+    PassthruQNode requires some changes to the way other PennyLane components work:
+
+    1. :class:`Operator` must not do domain checking for its parameters, or it must let the ADT pass the check.
+    2. The simulator device must return the result as the ADT instead of plain Python/NumPy types.
+    3. Any output_conversion in :meth:`BaseQNode.evaluate` must be skipped.
+
+    Additionally, any array-like ADT needs to be able to handle (1) scalar multiplication,
+    (2) indexing/slicing, and possibly (3) iteration, as these are the things qfuncs expect of
+    array-like parameters.
+    """
+    def __init__(self, func, device, properties=None):
+        super().__init__(func, device, mutable=True, properties=properties)
+
+    def __repr__(self):
+        """String representation."""
+        detail = "<PassthruQNode: device='{}', func={}, wires={}>"
+        return detail.format(self.device.short_name, self.func.__name__, self.num_wires)
+
+    def _set_variables(self, args, kwargs):
+        # do nothing, since we do not use VariableRefs
+        pass
+
+    def _construct(self, args, kwargs):
+        """Construct the quantum circuit graph by calling the quantum function.
+
+        Like :class:`.BaseQNode._construct`, but does not use VariableRefs.
+
+        PassthruQNode does not have a Jacobian method, so it does not HAVE to use VariableRefs or scalar linear indexing of input parameters.
+        Two options:
+        1. Use VariableRefs anyway, re-use most BaseQNode methods.
+           Problem: after evaluating the VariableRefs, stacking sliced/indexed Tensors in Operation.parameters should somehow result in a Tensor, not an object array.
+        2. Do not use VariableRefs, call the qfunc each time :meth:`PassthruQNode.evaluate` is called (always mutable).
+           Problem: tensornet_tf requires variable_deps?
+
+
+        TODO rethink output_conversion? should require device to return things in a fixed form, but either as arrays or as AD Tensors, do conversion in interface (if necessary...)
+        """
+        # temporary queues for operations and observables
+        self.queue = []  #: list[Operation]: applied operations
+        self.obs_queue = []  #: list[Observable]: applied observables
+
+        # set up the context for Operator entry
+        if qml._current_context is None:
+            qml._current_context = self
+        else:
+            raise QuantumFunctionError(
+                "qml._current_context must not be modified outside this method."
+            )
+        try:
+            # turn off domain checking since PassthruQNode qfuncs can take any class as input
+            pennylane.operation.Operator.do_check_domain = False
```

Comment:
Would it be possible to automatically disable the domain check if the parameters are "foreign", i.e. not `float` or `ndarray`?

Reply:
This would work, but it'd affect all types of QNodes and also let through unintended types passed as parameters by mistake.

Reply:
I have a couple of points here, let's see if any of these can help out at least a tiny bit :)
* "let through unintended types passed as parameters by mistake." - this part seems to be true for the cases captured by the checking right now.

* "PassthruQNode qfuncs can take any class as input" - as per my current understanding this is fixed upon execution, however, right? So in the `tensornet.tf` case it is for example `tf.tensor`.

If so, one approach could be the following:
1. Creating an `Operator.symbol_type` class attribute and in `Operator.__init__` setting it as `Operator.symbol_type = VariableRef`
2. Performing the first check in `check_domain` using `Operator.symbol_type` (currently line 375).
3. Upon initiating the `PassthruQnode`, the type of the variables can be determined by setting the  `Operator.symbol_type` class attribute
4. Once the execution of the `PassthruQnode` has concluded, this class attribute should be reset
(An alternative would be to set this attribute upon every `Qnode` instantiation, and set it with `Operator.symbol_type = VariableRef` for most cases.

One of the drawbacks of this approach would be related to parallel executions on simulators.

Another approach could be checking for `qml._current_context` and getting this information potentially from there (as the default `VariableRef` would need to be altered only once inside of particular `PassthruQnodes`)

Reply:
I admit using a global variable or class attribute to change the behavior of the library is not a very good practice, for various reasons (user errors, threading, code readability...).

Possible threading issues aside, maybe you could use a context manager to turn modify the domain checking behavior, so that when the context is exited the original behavior is automatically restored.

Reply:
Another issue is that for performance reasons we'd like to keep the domain checking reasonably light and quick, since it happens several times each time the QNode or its Jacobian is evaluated.

Reply:
Yes, I agree, perhaps this might actually require an ADR!

The checks could also be executed only when the QNode has been created, right?